File: 389-find-the-difference/findTheDifference.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from collections import defaultdict
class Solution(object):
    def findTheDifference(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        myDict = defaultdict(lambda: 0)
        for c in s:
            myDict[c] -= 1
        for c in t:
            myDict[c] += 1
            if myDict[c] == 1:
                return c
        return 0

s = Solution()
print(s.findTheDifference('abcd', 'abcde'))

File: 389-find-the-difference/findTheDifference.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Solution {
public:
    /**
        解法1，map
    */
    char findTheDifference(string s, string t) {
        unordered_map<char, int> myMap;
        for (auto c : s) myMap[c]--;
        for (auto c : t) {
            myMap[c]++;
            if (myMap[c] == 1) return c;
        }
        return 'a';
    }
    /**
        解法2，位运算
        char 如果不初始化的话值应该是null,Ascii是0？怎么不太对
    */
    char findTheDifference(string s, string t) {
        char ret = 0;
        for (auto c : s) ret ^= c;
        for (auto c : t) ret ^= c;
        return ret;
    }

};

int main() {
    Solution s =  Solution();
    cout << s.findTheDifference("abcd", "abcde") << endl;
    return 0;
}
